### PR TITLE
Bug 925988 - [B2G] [Contacts] Facebook contacts are present when choosin...

### PR DESCRIPTION
**Pull request by adrm**
...g which contacts to export

### CODE REVIEW

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=0795e03601ec)

File: apps/communications/contacts/js/views/list.js [hunk 1]
```diff
@@ -1731,7 +1734,18 @@ contacts.List = (function() {
         selectAllPending = true && !loaded;
         selectAllContacts();
 
-        currentlySelected = contacts.List.total;
+        var filtered;
+        if (currentFilterClass) {
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/list.js]: if(Array.isArray(currentFilterClass)

File: apps/communications/contacts/js/views/list.js [hunk 2]
```diff
@@ -1731,7 +1734,18 @@ contacts.List = (function() {
...
+          currentFilterClass.forEach(function(filterClass) {
+            switch (filterClass) {
+              case DISABLE_FB_CONTACTS_CLASS:
+                filtered = document.querySelectorAll('li[data-fb-uid]').length;
+                break;
+            }
+          });
+        }
+
+        currentlySelected = contacts.List.total - filtered||0;
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/list.js]: || 0 sentence is not needed

File: apps/communications/contacts/js/views/list.js [hunk 3]
```diff
@@ -1731,7 +1734,18 @@ contacts.List = (function() {
         selectAllPending = true && !loaded;
         selectAllContacts();
 
-        currentlySelected = contacts.List.total;
+        var filtered;
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/list.js]: filtered = 0

File: apps/communications/contacts/js/views/list.js [hunk 4]
```diff
@@ -50,6 +50,7 @@ contacts.List = (function() {
       groupList = null,
       searchList = null,
       currentlySelected = 0,
+      currentFilterClass = null,
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/list.js]: Better to name this variable currentFilterClassList

File: apps/communications/contacts/js/views/list.js [hunk 5]
```diff
@@ -1879,6 +1893,14 @@ contacts.List = (function() {
     }
     searchList.classList.add('selecting');
 
+    if (options.filterClassList) {
+      currentFilterClass = options.filterClassList;
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/list.js]: currentFilterClassList

File: apps/communications/contacts/style/app.css [hunk 6]
```diff
@@ -552,6 +552,16 @@ nav[data-type="scrollbar"] ol > li > * {
   display: block !important;
 }
 
+.selecting.disable-fb li[data-fb-uid] {
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/style/app.css]: the '.disable-fb' class should be independent of the 'selecting' class

Comment (jmcanterafonseca) [inline apps/communications/contacts/style/app.css]: bettter to name this class disable-fb-items and put a comment indicating its purpose

File: apps/communications/contacts/style/app.css [hunk 7]
```diff
@@ -552,6 +552,16 @@ nav[data-type="scrollbar"] ol > li > * {
...
+  color: #333333;
+  opacity: 0.25;
+  pointer-events: none;
+}
+
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/style/app.css]: please add a comment indicating the purpose of this rule

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=0997686bb0f6)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=8029ccb97bcc)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=74855b5bff5a)

File: apps/communications/contacts/js/views/settings.js [hunk 8]
```diff
@@ -373,7 +373,13 @@ contacts.Settings = (function() {
       navigationHandler,
       {
         isDanger: false,
-        transitionLevel: EXPORT_TRANSITION_LEVEL
+        transitionLevel: EXPORT_TRANSITION_LEVEL,
+        filterList: [
+          {
+            'class': 'disable-fb-items',
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/settings.js]: containerClass

File: apps/communications/contacts/js/views/settings.js [hunk 9]
```diff
@@ -373,7 +373,13 @@ contacts.Settings = (function() {
...
+            'selector': 'li[data-fb-uid]'
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/settings.js]: itemSelector

File: apps/communications/contacts/js/views/list.js [hunk 10]
```diff
@@ -2063,6 +2080,14 @@ contacts.List = (function() {
     fastScroll.classList.remove('selecting');
     utils.alphaScroll.toggleFormat('normal');
 
+    if (currentFilterList && Array.isArray(currentFilterList)){
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/list.js]: the condition is redundant

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=0717dea4ebba)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=6ca3cc26668b)

File: apps/communications/contacts/js/views/list.js [hunk 11]
```diff
@@ -1907,6 +1924,19 @@ contacts.List = (function() {
     identify such action.
 
     Also provide a callback to be invoked when we enter in select mode.
+
+    The 'options' object allows filtering of contacts passing an array of
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/list.js]: nit: allows to filter out contact

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=a53b322760c9)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=995d16db2541)

File: apps/communications/contacts/js/views/settings.js [hunk 12]
```diff
@@ -346,36 +346,55 @@ contacts.Settings = (function() {
...
+        },
+        null,
+        navigationHandler,
+        {
+          isDanger: false,
+          transitionLevel: EXPORT_TRANSITION_LEVEL,
+          filterList: [
+            {
+              'containerClass': 'disable-fb-items',
+              'filteredContacts': filteredContacts
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/settings.js]: numFilteredContacts

File: apps/communications/contacts/js/views/settings.js [hunk 13]
```diff
@@ -346,36 +346,55 @@ contacts.Settings = (function() {
...
-      }
-    );
+
+    var req = fb.utils.getNumFbContacts();
+
+    req.onsuccess = function() {
+      openSelectList(req.result);
+    };
+
+    req.onerror = function() {
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/settings.js]: console.error

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=166cbb5121da)

File: apps/communications/contacts/js/views/settings.js [hunk 14]
```diff
@@ -346,36 +346,56 @@ contacts.Settings = (function() {
...
+      console.error('fb.utils.getNumFbContacts() failed.', req.error.name);
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/settings.js]: nit: req.error && req.error.name

File: apps/communications/contacts/js/views/settings.js [hunk 15]
```diff
@@ -346,36 +346,56 @@ contacts.Settings = (function() {
...
-        });
-      },
-      null,
-      navigationHandler,
-      {
-        isDanger: false,
-        transitionLevel: EXPORT_TRANSITION_LEVEL
-      }
-    );
+
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/settings.js]: A comment is needed here in order to document why we are trying here to get the number of FB Contacts

File: apps/communications/contacts/js/views/settings.js [hunk 16]
```diff
@@ -967,10 +994,28 @@ contacts.Settings = (function() {
...
+            exportButton.setAttribute('disabled', 'disabled');
+            bulkDeleteButton.removeAttribute('disabled');
+            setICEButton.setAttribute('disabled', 'disabled');
+          } else {
+            enableButtons();
+          }
+        };
+
+        fbReq.onerror = function() {
+          window.console.warn('Number of fb contacts couldn\'t be retrieved',
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/settings.js]: nit console.error,  could not

File: apps/communications/contacts/js/views/settings.js [hunk 17]
```diff
@@ -967,10 +994,28 @@ contacts.Settings = (function() {
         bulkDeleteButton.setAttribute('disabled', 'disabled');
         setICEButton.setAttribute('disabled', 'disabled');
       }
-      else {
-         exportButton.removeAttribute('disabled');
-         bulkDeleteButton.removeAttribute('disabled');
-         setICEButton.removeAttribute('disabled');
+      else if (fb.isLinked) {
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/settings.js]: why are you calling fb.isLinked here?

File: apps/communications/contacts/js/views/settings.js [hunk 18]
```diff
@@ -960,6 +980,13 @@ contacts.Settings = (function() {
...
   var checkNoContacts = function checkNoContacts() {
     var exportButton = exportContacts.firstElementChild;
+
+    var enableButtons = function() {
+      exportButton.removeAttribute('disabled');
+      bulkDeleteButton.removeAttribute('disabled');
+      setICEButton.removeAttribute('disabled');
+    };
+
     var req = navigator.mozContacts.getCount();
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/js/views/settings.js]: I would suggest we do something like

var operations = [
  getNumContacts(),
  getNumFbContacts()
]

Promise.all(operations).then(function(results) {
})

and taking into account the results we enable disable the export, bulk and setICE buttons

**Comment by jmcanterafonseca** [comment]
We need additional tests for the disabled buttons regarding Facebook

File: apps/communications/contacts/style/app.css [hunk 19]
```diff
@@ -556,6 +556,17 @@ nav[data-type="scrollbar"] ol > li > * {
...
 }
 
+/* FB contacts will be disabled when adding .disable-fb-items to contact list */
+.disable-fb-items li[data-fb-uid] {
+  color: #333333;
+  opacity: 0.25;
+  pointer-events: none;
+}
+
+.disable-fb-items li[data-fb-uid] label span:after{
```
Comment (jmcanterafonseca) [inline apps/communications/contacts/style/app.css]: nit: blank space after the span:after

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=27f68db671c6)

**Comment by try-server-hook** [comment]
[ADLR-es Adrián de la Rosa (ADLR-es)](https://github.com/ADLR-es) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=00f09b2c37fc)

**Pull request merged**